include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+   */
+  std::function<int(const VectorType &src, VectorType &dst)> residual;
+
+  /**
+   * A function object that users may supply and that is intended to
+   * prepare the linear solver for subsequent calls to
+   * solve_jacobian_system().
+   *
+   * The job of setup_jacobian() is to prepare the linear solver for
+   * subsequent calls to solve_with_jacobian(), in the solution of linear
+   * systems $Ax = b$. The exact nature of this system depends on the
+   * SolutionStrategy that has been selected.
+   *
+   * In the cases strategy = SolutionStrategy::newton or
+   * SolutionStrategy::linesearch, $A$ is the Jacobian $J = \partial
+   * F/\partial u$. If strategy = SolutionStrategy::picard, $A$ is the
+   * approximate Jacobian matrix $L$.
+   *
+   * The setup_jacobian() function may call a user-supplied function, or a
+   * function within the linear solver module, to compute Jacobian-related
+   * data that is required by the linear solver. It may also preprocess that
+   * data as needed for solve_with_jacobian(), which may involve calling a
+   * generic function (such as for LU factorization) or, more generally,
+   * build preconditioners from the assembled Jacobian. In any case, the
+   * data so generated may then be used whenever a linear system is solved.
+   *
+   * @param current_u Current value of $u$
+   */
+  std::function<int(const VectorType &current_u)> setup_jacobian;
+
+  /**
+   * A function object that users may supply and that is intended to solve
+   * a linear system with the Jacobian matrix.
+   *
+   * Specific details on this function can be found in the KINSOL
+   *
+   * Arguments to the function are:
+   *
+   * @param[in] rhs The system right hand side to solve for.
+   * @param[out] dst The solution of $J^{-1} * src$.
+   * @param[in] tolerance The tolerance with which to solve the linear system
+   *   of equations.
+   */
+  std::function<
+    int(const VectorType &rhs, VectorType &dst, const double tolerance)>
+    solve_with_jacobian;
+
+protected:
+  /**
+   * NonlinearSolverSelector configuration data.
+   */
+  AdditionalData additional_data;
+
+private:
+  /**
+   * The MPI communicator to be used by this solver, if any.
+   */
+  MPI_Comm mpi_communicator;
+
+/**
+ * KINSOL configuration data
+ */
+#ifdef DEAL_II_WITH_SUNDIALS
+  typename SUNDIALS::KINSOL<VectorType>::AdditionalData additional_data_kinsol;
+#endif
+
+/**
+ * NOX configuration data
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+  typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+                                       additional_data_nox;
+  Teuchos::RCP<Teuchos::ParameterList> parameters_nox =
+    Teuchos::rcp(new Teuchos::ParameterList);
+#endif
+
+/**
+ * PETSc SNES configuration data
+ */
+#ifdef DEAL_II_WITH_PETSC
+  typename PETScWrappers::NonlinearSolverData additional_data_petsc_snes;
+#endif
+
+  /**
+   * Data transfer function
+   */
+  void
+  data_transfer(const AdditionalData &additional_data);
```

Comment:
Can we give this function a better name and description? It's not clear to me what kind of data is being transferred.